src/base_obs.c

```diff
@@ -0,0 +1,329 @@
+  chMtxUnlock();
+  /* Signal that a complete base observation has been received. */
+  chBSemSignal(&base_obs_received);
+}
+
+/** SBP callback for observation messages.
+ * A set of observations may be split across several messages so this function
+ * attempts to collect a full set of observations into a single `obss_t`
+ * (`base_obss_raw`). Once a full set is received then update_obss() is called.
+ */
+static void obs_callback(u16 sender_id, u8 len, u8 msg[], void* context)
+{
+  (void) context;
+
+  static s16 prev_count = 0;
+  static gps_time_t prev_t = {.tow = 0.0, .wn = 0};
+  /* Using an extra obss_t so we don't need to overwrite the old set (used for
+   * low latency) when we end up with a bad new set. */
+  static obss_t base_obss_raw = {.has_pos = 0};
+
+  /* Sender ID of zero means that the messages are relayed observations,
+   * ignore them. */
+  if (sender_id == 0)
+    return;
+
+  /* Relay observations using sender_id = 0. */
+  sbp_send_msg_(MSG_PACKED_OBS, len, msg, 0);
+
+  gps_time_t t;
+  u8 total;
+  u8 count;
+
+  /* Decode the message header to get the time and how far through the sequence
+   * we are. */
+  unpack_obs_header((msg_obs_header_t*)msg, &t, &total, &count);
+
+  /* Check to see if the observation is aligned with our internal observations,
+   * i.e. is it going to time match one of our local obs. */
+  double epoch_count = t.tow * (soln_freq / obs_output_divisor);
+  if (fabs(epoch_count - round(epoch_count)) > TIME_MATCH_THRESHOLD) {
+    printf("Unaligned observation from base station ignored.\n");
+    return;
+  }
+
+  /* Calculate packet latency. */
+  if (time_quality >= TIME_COARSE) {
+    gps_time_t now = get_current_time();
+    float latency_ms = (float) ((now.tow - t.tow) * 1000.0);
+    log_obs_latency(latency_ms);
+  }
+
+  /* Verify sequence integrity */
+  if (count == 0) {
+    prev_t = t;
+    prev_count = 0;
+  } else if (prev_t.tow != t.tow ||
+             prev_t.wn != t.wn ||
+             prev_count + 1 != count) {
+    printf("Dropped one of the observation packets! Skipping this sequence.\n");
+    prev_count = -1;
+    return;
+  } else {
+    prev_count = count;
+  }
+
+  /* Calculate the number of observations in this message by looking at the SBP
+   * `len` field. */
+  u8 obs_in_msg = (len - sizeof(msg_obs_header_t)) / sizeof(msg_obs_content_t);
+
+  /* If this is the first packet in the sequence then reset the base_obss_raw
+   * state. */
+  if (count == 0) {
+    base_obss_raw.n = 0;
+    base_obss_raw.t = t;
+  }
+
+  /* Pull out the contents of the message. */
+  msg_obs_content_t *obs = (msg_obs_content_t *)(msg + sizeof(msg_obs_header_t));
+  for (u8 i=0; i<obs_in_msg; i++) {
+    /* Check if we have an ephemeris for this satellite, we will need this to
+     * fill in satellite position etc. parameters. */
+    if (ephemeris_good(es[obs[i].prn], t)) {
+      /* Unpack the observation into a navigation_measurement_t. */
+      unpack_obs_content(
```

Comment:
Because Niels did it that way initially, I guess it makes it more general purpose and less libswiftnav specific. I could change it but it might be out of the scope of this PR?